### PR TITLE
[JENKINS-23784] - Register `DependencyClassLoader` in the default plugin strategy as parallel-capable

### PR DESCRIPTION
**Pull request by basil**
`DependencyClassLoader` can be registered as parallel-capable with a modest amount of effort. First, it needs to be made static so that we can invoke `registerAsParallelCapable();`. Then we need to do an audit of this class for thread-safety.

The `private List<Dependency> dependencies` field is only ever iterated over, never mutated, and the caller always passes in a freshly-constructed list and doesn't do anything else with it, so we don't need to do anything here. But we do a defensive copy into an unmodifiable list for extra protection in case the caller changes in the future.

The `private volatile List<PluginWrapper> transientDependencies` field does require synchronization, but [double-checked locking](https://errorprone.info/bugpattern/DoubleCheckedLocking) solves that problem.

The tricky part is the use of `PluginManager`. We were originally reading this field from the outer class, but now that the inner `DependencyClassLoader` class is static we've got to pass it in as a constructor argument. The field is fortunately effectively final in the outer class, so we can just mark it `final` in both the outer class and inner class and not worry about it changing from underneath us.

But what about accessing `PluginManager`? We only ever call `getPlugin()` on it to get a `PluginWrapper`, and the underlying datastructure in `PluginManager` is a `CopyOnWriteArrayList`, so the iteration should be safe.

But what about accessing `PluginWrapper`? A list in `PluginWrapper` is modified by `resolvePluginDependencies()`, but that is only ever called before `PluginStrategy#load`, which is the first time the `DependencyClassLoader` is used. Similarly, `hasCycleDependency` is also set before `PluginStrategy#load`. The other fields in `PluginWrapper` all look safe.

And with that the audit is complete. In theory this should work. A test with `java -jar jenkins.war` and about 150 plugins led to a successful startup. I haven't done any stress testing beyond that.

### Proposed changelog entries

* Load classes from plugins in parallel for faster startup on multicore machines.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by oleg-nenashev — APPROVED** [review]
Looks good to me. This code is isolated to the `ClassicPluginStrategy` so other implementations should not be affected. Looking forward to seeing it in production

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by jglick — APPROVED** [review]
Sounds great. This is not “internal”, it is at least a partial fix of https://issues.jenkins.io/browse/JENKINS-23784 (does not deal with `UberClassLoader` or `PluginFirstClassLoader`) and should be marked as such.

**Comment by basil** [comment]
> This is not “internal”, it is at least a partial fix of https://issues.jenkins.io/browse/JENKINS-23784 (does not deal with `UberClassLoader` or `PluginFirstClassLoader`) and should be marked as such.

Sure. I updated the description to remove "internal" and reference JENKINS-23784. As you note, it is a partial fix. The last part is updating `UberClassLoader`, a change which I have prepared but not yet posted the PR for.

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (basil) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:580]: Indeed, I think you would always wind up calculating it. I do not know why Kohsuke opted for lazy initialization here, nor do I know why he called it "transient dependencies" (I think it should read "transitive"). Anyway, there is a lot of code that could be improved here, but as usual I am trying to prevent code sprawl by minimizing the number of unrelated changes in these PRs. If we are to minimize code change and continue the existing lazy initialization paradigm, then it ought to at least perform well in a concurrent environment by using a rigorous version of double-checked locking.

Comment (jglick) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:580]: Sure, makes sense, just something that caught my eye during the review that could perhaps be cleaned up later.

**Comment by oleg-nenashev** [comment]
I suggest "Allow parallel classloading in dependency classloader when using the default plugin strategy" as a changelog entry @basil

**Comment by jglick** [comment]
Or more comprehensibly (and omitting mention of plugin strategy which is not _really_ pluggable in practice)

> Load classes from plugins in parallel for faster startup on multicore machines

?

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (jtnord) [inline core/src/main/java/hudson/ClassicPluginStrategy.java:601]: NIT: as mentioned above more likely this should have been `localTransitiveDependencies` as this is new code.

**Review by jtnord — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
> plugin strategy which is not really pluggable in practice

Huh, why? The Entire Plugin Manager implementation can be replaced. I would prefer to be explicit there.

**Comment by jglick** [comment]
> The Entire Plugin Manager implementation can be replaced.

With what? By whom? For what purpose? I am aware of https://github.com/jenkinsci/jenkins-test-harness/blob/1130c213d1953e1356ce2e8d2c3e131b5da2a5ac/src/main/java/org/jvnet/hudson/test/TestPluginManager.java#L23 and https://github.com/jenkinsci/jenkinsfile-runner/blob/253701e2db085d40929e98f73fe43c4a1d4c816a/setup/src/main/java/io/jenkins/jenkinsfile/runner/PluginManagerImpl.java#L13 but nothing a Jenkins user would ever do.

**Pull request merged**
